### PR TITLE
Store listener lists inside the EventBus

### PR DESCRIPTION
**Pull request by Technici4n**
Benchmarks shows similar if not better performance:

```
Benchmark                                              Mode  Cnt   Score    Error  Units
EventBusBenchmark.testClassLoaderCombined              avgt   15  73.054 ± 22.724  ns/op
EventBusBenchmark.testClassLoaderCombined:·stack       avgt          NaN             ---
EventBusBenchmark.testClassLoaderDynamic               avgt   15  30.681 ±  0.873  ns/op
EventBusBenchmark.testClassLoaderDynamic:·stack        avgt          NaN             ---
EventBusBenchmark.testClassLoaderLambda                avgt   15  32.367 ±  2.214  ns/op
EventBusBenchmark.testClassLoaderLambda:·stack         avgt          NaN             ---
EventBusBenchmark.testClassLoaderStatic                avgt   15  30.204 ±  2.599  ns/op
EventBusBenchmark.testClassLoaderStatic:·stack         avgt          NaN             ---
EventBusBenchmark.testModLauncherCombined              avgt   15  64.087 ±  1.757  ns/op
EventBusBenchmark.testModLauncherCombined:·stack       avgt          NaN             ---
EventBusBenchmark.testModLauncherDynamic               avgt   15  31.168 ±  0.494  ns/op
EventBusBenchmark.testModLauncherDynamic:·stack        avgt          NaN             ---
EventBusBenchmark.testModLauncherLambda                avgt   15  31.201 ±  0.565  ns/op
EventBusBenchmark.testModLauncherLambda:·stack         avgt          NaN             ---
EventBusBenchmark.testModLauncherStatic                avgt   15  28.732 ±  1.052  ns/op
EventBusBenchmark.testModLauncherStatic:·stack         avgt          NaN             ---
EventBusBenchmarkNoLoader.testNoLoaderCombined         avgt   15  65.787 ±  1.878  ns/op
EventBusBenchmarkNoLoader.testNoLoaderCombined:·stack  avgt          NaN             ---
EventBusBenchmarkNoLoader.testNoLoaderDynamic          avgt   15  31.424 ±  4.117  ns/op
EventBusBenchmarkNoLoader.testNoLoaderDynamic:·stack   avgt          NaN             ---
EventBusBenchmarkNoLoader.testNoLoaderLambda           avgt   15  32.635 ±  4.949  ns/op
EventBusBenchmarkNoLoader.testNoLoaderLambda:·stack    avgt          NaN             ---
EventBusBenchmarkNoLoader.testNoLoaderStatic           avgt   15  29.688 ±  3.934  ns/op
EventBusBenchmarkNoLoader.testNoLoaderStatic:·stack    avgt          NaN             ---
```

### CODE REVIEW

File: src/main/java/net/neoforged/bus/ListenerList.java [hunk 1]
```diff
@@ -30,247 +29,109 @@
...
+            ret.addAll(parent.getListeners(priority));
+        }
+        return ret;
     }
 
-    protected ListenerListInst getInstance(int id)
-    {
-        return lists[id];
+    /**
+     * Returns a full list of all listeners for all priority levels.
```
Comment (Matyrobbrt) [inline src/main/java/net/neoforged/bus/ListenerList.java:76]:
```suggestion
     * {@return a full list of all listeners for all priority levels}
```
and remove the @return declaration.

File: src/main/java/net/neoforged/bus/ListenerList.java [hunk 2]
```diff
@@ -30,247 +29,109 @@
...
+    public IEventListener[] getListeners() {
+        if (shouldRebuild()) buildCache();
+        return listeners.get();
     }
 
-    public IEventListener[] getListeners(int id)
-    {
-        return lists[id].getListeners();
+    protected boolean shouldRebuild() {
+        return rebuild;// || (parent != null && parent.shouldRebuild());
```
Comment (Matyrobbrt) [inline src/main/java/net/neoforged/bus/ListenerList.java:91]: Why is this code commented out? If it's intentional, you should remove the comment.

**Review by Matyrobbrt — CHANGES_REQUESTED** [review]

File: src/main/java/net/neoforged/bus/ListenerList.java (hunk 2 above)
Comment (Technici4n) [reply src/main/java/net/neoforged/bus/ListenerList.java:91]: That was already the case before, I just moved the code.

File: src/main/java/net/neoforged/bus/ListenerList.java (hunk 1 above)
Comment (Technici4n) [reply src/main/java/net/neoforged/bus/ListenerList.java:76]: I can change this if you want, but I just moved the code that existed before to a different indentation level.

**Review by Matyrobbrt — APPROVED** [review]

**Pull request merged**